src/main/java/org/apache/lucene/BoundedLinearFeatureQuery.java

```diff
@@ -0,0 +1,235 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+/*
+ * This class is built based on lucene FeatureQuery. We use LinearFuntion to
+ * build the query and add an upperbound to it.
+ */
+
+package org.apache.lucene;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import org.apache.lucene.index.ImpactsEnum;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.PostingsEnum;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.ImpactsDISI;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.TermQuery;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.search.similarities.Similarity.SimScorer;
+import org.apache.lucene.util.BytesRef;
+
+/**
+ * The feature queries of input tokens are wrapped by lucene BooleanQuery, which use WAND algorithm
+ * to accelerate the execution. The WAND algorithm leverage the score upper bound of sub-queries to
+ * skip non-competitive tokens. However, origin lucene FeatureQuery use Float.MAX_VALUE as the score
+ * upper bound, and this invalidates WAND.
+ *
+ * To mitigate this issue, we rewrite the FeatureQuery to BoundedLinearFeatureQuery. The caller can
+ * set the token score upperbound of this query. And according to our use case, we use LinearFunction
+ * as the score function.
+ *
+ * This class combines both <a href="https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/document/FeatureQuery.java">FeatureQuery</a>
+ * and <a href="https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/document/FeatureField.java">FeatureField</a> together
+ * and will be deprecated after OpenSearch upgraded lucene to version 9.8.
+ */
+
+public final class BoundedLinearFeatureQuery extends Query {
```

Comment:
There are no unit test for this class. Can we add the UTs for this class.

Reply:
Also, what methods are different from the Lucene FeatureQuery implementation?

Reply:
This method is the core different: https://github.com/opensearch-project/neural-search/pull/348/files/ddb5d69d16b88da62f864146d9d29252c2d3773e#diff-fa6b12823de991fcb735f2da7e1ad61e3b58829a4291bcfa276e2a1d685b16a6R227.